File: Final_Project/Code/Project_Interger_Partition/Bai_toan_1/CSS_bt1.py
def generate_partitions(n, k, max_val, current, partitions):
    if k == 0:
        if n == 0:
            partitions.append(current[:])
        return
    if n < k:
        return
    for i in range(1, min(n - k + 1, max_val) + 1):
        current.append(i)
        generate_partitions(n - i, k - 1, i, current, partitions)
        current.pop()

def print_ferrers(partition):
    print("Ferrers diagram:")
    for val in partition:
        print("*" * val)

def print_ferrers_transpose(partition):
    print("Ferrers transpose diagram:")
    max_val = max(partition)
    transpose = [sum(1 for x in partition if x >= i) for i in range(1, max_val + 1)]
    for val in transpose:
        print("*" * val)

def main():
    n = int(input("Nhap n: "))
    k = int(input("Nhap k: "))
    partitions = []
    generate_partitions(n, k, n, [], partitions)
    for i, partition in enumerate(partitions):
        print(f"\nPartition {i + 1}: {partition}")
        print_ferrers(partition)
        print_ferrers_transpose(partition)

if __name__ == "__main__":
    main()

File: Final_Project/Code/Project_Interger_Partition/Bai_toan_1/CSS_bt1.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void generate_partitions(int n, int k, int max_val, vector<int>& current, vector<vector<int>>& partitions) {
    if (k == 0) {
        if (n == 0) partitions.push_back(current);
        return;
    }
    if (n < k) return;
    for (int i = 1; i <= min(n - k + 1, max_val); i++) {
        current.push_back(i);
        generate_partitions(n - i, k - 1, i, current, partitions);
        current.pop_back();
    }
}

void print_ferrers(const vector<int>& partition) {
    cout << "Ferrers diagram:" << endl;
    for (int val : partition) {
        for (int j = 0; j < val; j++) cout << "*";
        cout << endl;
    }
}

void print_ferrers_transpose(const vector<int>& partition) {
    cout << "Ferrers transpose diagram:" << endl;
    int max_val = *max_element(partition.begin(), partition.end());
    vector<int> transpose;
    for (int i = 1; i <= max_val; i++) {
        int count = 0;
        for (int val : partition) if (val >= i) count++;
        transpose.push_back(count);
    }
    for (int val : transpose) {
        for (int j = 0; j < val; j++) cout << "*";
        cout << endl;
    }
}

int main() {
    int n, k;
    cout << "Nhap n: "; cin >> n;
    cout << "Nhap k: "; cin >> k;
    vector<int> current;
    vector<vector<int>> partitions;
    generate_partitions(n, k, n, current, partitions);
    for (size_t i = 0; i < partitions.size(); i++) {
        cout << "\nPartition " << i + 1 << ": ";
        for (int val : partitions[i]) cout << val << " ";
        cout << endl;
        print_ferrers(partitions[i]);
        print_ferrers_transpose(partitions[i]);
    }
    return 0;
}